### PR TITLE
Add BeNaN and NotBeNaN assertions

### LINKED ISSUE #2579: Add support for asserting NaN values
### Background and motivation

It would be nice if there is an extension method for double and float values to check NaN values.

`
double value = double.NaN;
value.Should().BeNaN( );
`

### Alternative Concerns

_No response_

### Are you willing help with a pull-request?

No

### PR DESCRIPTION
**Pull request by arocheleau**
Add the `BeNaN` and `NotBeNaN` extension methods to `NumericAssertionsExtensions`.

```cs
float value = float.NaN;

value.Should().BeNaN();
value.Should().NotBeNaN();
```

This fixes #2579.

## IMPORTANT 

* [x] If the PR touches the public API, the changes have been approved in a separate issue with the "api-approved" label.
* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [x] The changes are covered by unit tests which follow the Arrange-Act-Assert syntax and the naming conventions such as is used [in these tests](../tree/develop/Tests/FluentAssertions.Equivalency.Specs/MemberMatchingSpecs.cs#L51-L430).
* [x] If the PR adds a feature or fixes a bug, please update [the release notes](../tree/develop/docs/_pages/releases.md) with a functional description that explains what the change means to consumers of this library, which are published on the [website](https://fluentassertions.com/releases).
* [x] If the PR changes the public API the changes needs to be included by running [AcceptApiChanges.ps1](../tree/develop/AcceptApiChanges.ps1) or [AcceptApiChanges.sh](../tree/develop/AcceptApiChanges.sh).
* [x] If the PR affects [the documentation](../tree/develop/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com/introduction).
    * [x] Please also run `./build.sh --target spellcheck` or `.\build.ps1 --target spellcheck` before pushing and check the good outcome

### CODE REVIEW

File: Tests/FluentAssertions.Specs/Numeric/NumericAssertionSpecs.BeNaN.cs [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+        public void When_a_float_value_is_NaN_it_should_succeed()
+        {
+            // Arrange
+            float actual = float.NaN;
+
+            // Act
+            Action act = () => actual.Should().BeNaN();
+
+            // Assert
+            act.Should().NotThrow();
```
Comment (IT-VBFK) [inline Tests/FluentAssertions.Specs/Numeric/NumericAssertionSpecs.BeNaN.cs:21]:
```suggestion
            // Act / Assert
            actual.Should().BeNaN();
```

We now combine assertions that are not supposed to throw into one.

(And following)

File: docs/_pages/releases.md [hunk 2]
```diff
@@ -7,14 +7,15 @@ sidebar:
...
-## 7.0 Alpha 1
+## 7.0 Alpha X
 
 ### What's new
 
 * Introduced a new assembly-level attribute that you can use to initialize Fluent Assertions before the first assertion - [#2292](https://github.com/fluentassertions/fluentassertions/pull/2292)
 * Ensure compatibility with .NET 8 - [#2466](https://github.com/fluentassertions/fluentassertions/pull/2466)
 * Add support for NUnit 4 - [#2483](https://github.com/fluentassertions/fluentassertions/pull/2483)
 * Added `NotBeIn` to check if a `DateTime` is not in a given `DateTimeKind` - [#2536](https://github.com/fluentassertions/fluentassertions/pull/2536)
+* Added `BeNaN` and `NotBeNaN` for assertions on `float` and `double` - [#2579](https://github.com/fluentassertions/fluentassertions/pull/2579)
```
Comment (jnyrup) [inline docs/_pages/releases.md:18]: We've so far been referencing the PR with the changes

```suggestion
* Added `BeNaN` and `NotBeNaN` for assertions on `float` and `double` - [#2606](https://github.com/fluentassertions/fluentassertions/pull/2606)
```

File: Src/FluentAssertions/NumericAssertionsExtensions.cs [hunk 3]
```diff
@@ -1481,6 +1481,209 @@ public static class NumericAssertionsExtensions
 
     #endregion
 
+    #region BeNaN
+
+    /// <summary>
+    /// Asserts that the number value is NaN.
```
Comment (jnyrup) [inline Src/FluentAssertions/NumericAssertionsExtensions.cs:1487]:
```suggestion
    /// Asserts that the float value is NaN.
```

File: Src/FluentAssertions/NumericAssertionsExtensions.cs [hunk 4]
```diff
@@ -1481,6 +1481,209 @@ public static class NumericAssertionsExtensions
...
+    {
+        float actualValue = parent.Subject.Value;
+
+        FailIfValueIsNotNaN(float.IsNaN(actualValue), actualValue, because, becauseArgs);
+
+        return new AndConstraint<NumericAssertions<float>>(parent);
+    }
+
+    /// <summary>
+    /// Asserts that the number value is NaN.
```
Comment (jnyrup) [inline Src/FluentAssertions/NumericAssertionsExtensions.cs:1509]:
```suggestion
    /// Asserts that the double value is NaN.
```

File: Tests/FluentAssertions.Specs/Numeric/NumericAssertionSpecs.BeNaN.cs [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+using Xunit.Sdk;
+
+namespace FluentAssertions.Specs.Numeric;
+
+public partial class NumericAssertionSpecs
+{
+    public class BeNaN
+    {
+        [Fact]
+        public void When_a_float_value_is_NaN_it_should_succeed()
```
Comment (jnyrup) [inline Tests/FluentAssertions.Specs/Numeric/NumericAssertionSpecs.BeNaN.cs:12]: We have been moving away from the `When_[..]_it_should_[..]` naming scheme and over to a less strict naming that focuses more even more on the behavior being tested and less about fitting into to some scheme.
E.g. when most tests end with `it_should_succeed` that suffix provides little to no value.

See https://enterprisecraftsmanship.com/posts/you-naming-tests-wrong/

File: Src/FluentAssertions/NumericAssertionsExtensions.cs [hunk 6]
```diff
@@ -1481,6 +1481,209 @@ public static class NumericAssertionsExtensions
 
     #endregion
 
+    #region BeNaN
+
+    /// <summary>
+    /// Asserts that the float value is NaN.
```
Comment (dennisdoomen) [inline Src/FluentAssertions/NumericAssertionsExtensions.cs:1487]: 🤔 I prefer to keep this more functional, e.g. like `Asserts that a number is seen as not-a-number`.

File: Src/FluentAssertions/NumericAssertionsExtensions.cs [hunk 7]
```diff
@@ -1481,6 +1481,209 @@ public static class NumericAssertionsExtensions
...
+        params object[] becauseArgs)
+    {
+        double? actualValue = parent.Subject;
+
+        FailIfValueIsNaN(actualValue is { } value && double.IsNaN(value), because, becauseArgs);
+
+        return new AndConstraint<NullableNumericAssertions<double>>(parent);
+    }
+
+    private static void FailIfValueIsNaN(bool valueIsNaN,
```
Comment (dennisdoomen) [inline Src/FluentAssertions/NumericAssertionsExtensions.cs:1676]: 🤔 I'm not a big fan of applying DRY like that. In most cases, duplicating the code makes the implementation much more readable.

**Review by dennisdoomen — APPROVED** [review]
It would be nice to update the `numerictypes.md` with the two new methods.

**Comment by arocheleau** [comment]
I did add them to the `numerictypes.md` earlier, but I guess I missunderstood your comment saying "🌱 I generally don't document these since they are not relevant for users.".

**Review by jnyrup — APPROVED** [review]

**Comment by dennisdoomen** [comment]
> I did add them to the `numerictypes.md` earlier, but I guess I missunderstood your comment saying "🌱 I generally don't document these since they are not relevant for users.".

I meant the `this parent` parameter ;-)

**Pull request merged**